### PR TITLE
optimize: refactor logic of pilot gateway to avoid extra iteration time

### PR DESCRIPTION
**Pull request by chlins**
Signed-off-by: chlins <chlins.zhang@gmail.com>

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ X] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 1]
```diff
@@ -108,10 +110,12 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(
 					// This is a HTTPS server, where we are doing TLS termination. Build a http connection manager with TLS context
 					routeName := mergedGateway.RouteNamesByServer[server]
 					filterChainOpts = append(filterChainOpts, configgen.createGatewayHTTPFilterChainOpts(node, server, routeName, push.Mesh.SdsUdsPath))
+					filterChains = append(filterChains, plugin.FilterChain{ListenerProtocol: plugin.ListenerProtocolHTTP})
 				} else {
 					// passthrough or tcp, yields multiple filter chains
 					filterChainOpts = append(filterChainOpts, configgen.createGatewayTCPFilterChainOpts(node, push,
 						server, map[string]bool{mergedGateway.GatewayNameForServer[server]: true})...)
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/gateway.go]: Actually `createGatewayTCPFilterChainOpts` return value is of array type. It is confusing as you do this below, so could you also change that function return value?

Comment (chlins) [reply pilot/pkg/networking/core/v1alpha3/gateway.go]: @hzxuzhonghu Thank you for your review, i'm sorry to forget change the return value of the createGatewayTCPFilterChainOpts.

Comment (chlins) [reply pilot/pkg/networking/core/v1alpha3/gateway.go]: @hzxuzhonghu It's looks the function createGatewayTCPFilterChainOpts has some special logic to return array values when case of passthrough server to return buildGatewayNetworkFiltersFromTLSRoutes function, so i think we can handle array to what we needs.

**Pull request merged**